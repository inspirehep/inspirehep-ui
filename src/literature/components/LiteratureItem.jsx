import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Map } from 'immutable';

import AuthorList from './AuthorList';
import LiteratureDate from './LiteratureDate';
import ArxivEprintList from './ArxivEprintList';
import CollaborationList from './CollaborationList';
import DOIList from './DOIList';
import PartialAbstract from './PartialAbstract/PartialAbstract';
import ReportNumberList from './ReportNumberList';
import PublicationInfoList from './PublicationInfoList';
import ArxivPdfDownloadAction from './ArxivPdfDownloadAction';
import CiteModalAction from './CiteModalAction';
import ListItemAction from '../../common/components/ListItemAction';
import Latex from '../../common/components/Latex';
import ResultItem from '../../common/components/ResultItem';

class LiteratureItem extends Component {
  render() {
    const { metadata } = this.props;

    const title = metadata.getIn(['titles', 0, 'title']);
    const authors = metadata.get('authors');

    const arxivId = metadata.getIn(['arxiv_eprints', 0, 'value']);
    const recordId = metadata.get('control_number');
    const citationCount = metadata.get('citation_count');
    const referenceCount = metadata.get('number_of_references');
    const authorCount = metadata.get('number_of_authors');

    const date = metadata.get('date');
    const publicationInfo = metadata.get('publication_info');
    const eprints = metadata.get('arxiv_eprints');
    const dois = metadata.get('dois');
    const collaborations = metadata.get('collaborations');
    const reportNumbers = metadata.get('report_numbers');
    const abstract = metadata.getIn(['abstracts', 0, 'value']);

    return (
      <ResultItem
        actions={[
          arxivId && <ArxivPdfDownloadAction arxivId={arxivId} />,
          <CiteModalAction recordId={recordId} />,
          referenceCount && (
            <ListItemAction
              iconType="login"
              text={`${referenceCount} references`}
              link={{ to: `/literature/${recordId}#references` }}
            />
          ),
          citationCount && (
            <ListItemAction
              iconType="logout"
              text={`${citationCount} citations`}
              link={{ to: `/literature/${recordId}#citations` }}
            />
          ),
        ].filter(action => action != null)}
      >
        <Link className="f4" to={`/literature/${recordId}`}>
          <Latex>{title}</Latex>
        </Link>
        <div className="mt2">
          <div>
            <CollaborationList collaborations={collaborations} />
            <AuthorList
              authorCount={authorCount}
              recordId={recordId}
              authors={authors}
              limit={collaborations ? 1 : 5}
            />
          </div>
          <LiteratureDate date={date} />
        </div>
        <div className="mt2">
          <PublicationInfoList publicationInfo={publicationInfo} />
          <ArxivEprintList eprints={eprints} />
          <DOIList dois={dois} />
          <ReportNumberList reportNumbers={reportNumbers} />
          <PartialAbstract abstract={abstract} />
        </div>
      </ResultItem>
    );
  }
}

LiteratureItem.propTypes = {
  metadata: PropTypes.instanceOf(Map).isRequired,
};

export default LiteratureItem;
