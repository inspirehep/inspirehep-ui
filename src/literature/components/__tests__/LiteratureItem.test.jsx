import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import LiteratureItem from '../LiteratureItem';

describe('LiteratureItem', () => {
  it('renders with all props set, including sub props', () => {
    const metadata = fromJS({
      date: 'Jun 7, 1993',
      number_of_references: 2,
      titles: [{ title: 'test' }],
      authors: [{ full_name: 'Test, Author' }],
      arxiv_eprints: [{ value: '1234567890' }],
      control_number: 12345,
      citation_count: 1,
      publication_info: [{ journal_title: 'Test Jornal' }],
      dois: [{ value: '12345679.1234.123' }],
      report_numbers: [{ value: 'ABCD-AB-CD-1234-123' }],
      abstract: [{ value: 'Test Abstract' }],
      collaborations: [{ value: 'CMS' }],
      collaborations_with_suffix: [{ value: 'CMS Group' }],
    });
    const wrapper = shallow(<LiteratureItem metadata={metadata} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('does not arxiv pdf download action if there is no eprint value', () => {
    const metadata = fromJS({
      control_number: 12345,
    });
    const wrapper = shallow(<LiteratureItem metadata={metadata} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders 0 citations and 0 references if they do not exist', () => {
    const metadata = fromJS({
      control_number: 12345,
    });
    const wrapper = shallow(<LiteratureItem metadata={metadata} />);
    expect(wrapper).toMatchSnapshot();
  });
});
