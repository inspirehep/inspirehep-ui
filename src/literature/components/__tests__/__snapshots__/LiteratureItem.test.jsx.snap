// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiteratureItem does not arxiv pdf download action if there is no eprint value 1`] = `
<ResultItem
  actions={
    Array [
      <CiteModalAction
        recordId={12345}
    />,
    ]
  }
>
  <Link
    className="f4"
    replace={false}
    to="/literature/12345"
  >
    <Latex />
  </Link>
  <div
    className="mt2"
  >
    <AuthorList
      authors={
        Immutable.List [
        ]
      }
      limit={5}
      recordId={12345}
    />
    <LiteratureDate
      date={null}
    />
  </div>
  <div
    className="mt2"
  >
    <PublicationInfoList
      publicationInfo={null}
    />
    <ArxivEprintList
      eprints={null}
    />
    <DOIList
      dois={null}
    />
    <ReportNumberList
      reportNumbers={null}
    />
    <CollapsableAbstract
      abstract={null}
    />
  </div>
</ResultItem>
`;

exports[`LiteratureItem does not render citation and reference count if they are missing 1`] = `
<ResultItem
  actions={
    Array [
      <CiteModalAction
        recordId={12345}
    />,
    ]
  }
>
  <Link
    className="f4"
    replace={false}
    to="/literature/12345"
  >
    <Latex />
  </Link>
  <div
    className="mt2"
  >
    <AuthorList
      authors={
        Immutable.List [
        ]
      }
      limit={5}
      recordId={12345}
    />
    <LiteratureDate
      date={null}
    />
  </div>
  <div
    className="mt2"
  >
    <PublicationInfoList
      publicationInfo={null}
    />
    <ArxivEprintList
      eprints={null}
    />
    <DOIList
      dois={null}
    />
    <ReportNumberList
      reportNumbers={null}
    />
    <CollapsableAbstract
      abstract={null}
    />
  </div>
</ResultItem>
`;

exports[`LiteratureItem renders with all props set, including sub props 1`] = `
<ResultItem
  actions={
    Array [
      <ArxivPdfDownloadAction
        arxivId="1234567890"
    />,
      <CiteModalAction
        recordId={12345}
    />,
      <ListItemAction
        iconType="login"
        link={
            Object {
                "to": "/literature/12345#references",
              }
        }
        onClick={undefined}
        text="2 references"
    />,
      <ListItemAction
        iconType="logout"
        link={
            Object {
                "to": "/literature/12345#citations",
              }
        }
        onClick={undefined}
        text="1 citations"
    />,
    ]
  }
>
  <Link
    className="f4"
    replace={false}
    to="/literature/12345"
  >
    <Latex>
      test
    </Latex>
  </Link>
  <div
    className="mt2"
  >
    <AuthorList
      authors={
        Immutable.List [
          Immutable.Map {
            full_name: "Test, Author",
          },
        ]
      }
      limit={5}
      recordId={12345}
    />
    <LiteratureDate
      date="Jun 7, 1993"
    />
  </div>
  <div
    className="mt2"
  >
    <PublicationInfoList
      publicationInfo={null}
    />
    <ArxivEprintList
      eprints={
        Immutable.List [
          Immutable.Map {
            value: "1234567890",
          },
        ]
      }
    />
    <DOIList
      dois={
        Immutable.List [
          Immutable.Map {
            value: "12345679.1234.123",
          },
        ]
      }
    />
    <ReportNumberList
      reportNumbers={
        Immutable.List [
          Immutable.Map {
            value: "ABCD-AB-CD-1234-123",
          },
        ]
      }
    />
    <CollapsableAbstract
      abstract={null}
    />
  </div>
</ResultItem>
`;
